File: decompiled/sources/com/hjq/widget/layout/NestedScrollWebView.java
package com.hjq.widget.layout;

import android.content.Context;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.webkit.WebView;
import androidx.core.view.NestedScrollingChild;
import androidx.core.view.NestedScrollingChildHelper;

public class NestedScrollWebView extends WebView implements NestedScrollingChild {
    private boolean mChange;
    private final NestedScrollingChildHelper mChildHelper = new NestedScrollingChildHelper(this);
    private int mLastMotionY;
    private int mNestedOffsetY;
    private final int[] mScrollConsumed = new int[2];
    private final int[] mScrollOffset = new int[2];

    public NestedScrollWebView(Context context) {
        super(context);
        setNestedScrollingEnabled(true);
    }

    public NestedScrollWebView(Context context, AttributeSet attributeSet) {
        super(context, attributeSet);
        setNestedScrollingEnabled(true);
    }

    public NestedScrollWebView(Context context, AttributeSet attributeSet, int i) {
        super(context, attributeSet, i);
        setNestedScrollingEnabled(true);
    }

    public NestedScrollWebView(Context context, AttributeSet attributeSet, int i, int i2) {
        super(context, attributeSet, i, i2);
        setNestedScrollingEnabled(true);
    }

    public boolean onTouchEvent(MotionEvent motionEvent) {
        MotionEvent obtain = MotionEvent.obtain(motionEvent);
        int actionMasked = motionEvent.getActionMasked();
        boolean z = false;
        if (actionMasked == 0) {
            this.mNestedOffsetY = 0;
        }
        int y = (int) motionEvent.getY();
        motionEvent.offsetLocation(0.0f, (float) this.mNestedOffsetY);
        if (actionMasked != 0) {
            if (actionMasked != 1) {
                if (actionMasked == 2) {
                    int i = this.mLastMotionY - y;
                    if (dispatchNestedPreScroll(0, i, this.mScrollConsumed, this.mScrollOffset)) {
                        i -= this.mScrollConsumed[1];
                        obtain.offsetLocation(0.0f, (float) this.mScrollOffset[1]);
                        this.mNestedOffsetY += this.mScrollOffset[1];
                    }
                    this.mLastMotionY = y - this.mScrollOffset[1];
                    int scrollY = getScrollY();
                    int max = Math.max(0, scrollY + i) - scrollY;
                    if (dispatchNestedScroll(0, max, 0, i - max, this.mScrollOffset)) {
                        int i2 = this.mLastMotionY;
                        int[] iArr = this.mScrollOffset;
                        this.mLastMotionY = i2 - iArr[1];
                        obtain.offsetLocation(0.0f, (float) iArr[1]);
                        this.mNestedOffsetY += this.mScrollOffset[1];
                    }
                    if (this.mScrollConsumed[1] == 0 && this.mScrollOffset[1] == 0) {
                        if (this.mChange) {
                            this.mChange = false;
                            obtain.setAction(0);
                            super.onTouchEvent(obtain);
                        } else {
                            z = super.onTouchEvent(obtain);
                        }
                        obtain.recycle();
                        return z;
                    } else if (this.mChange) {
                        return false;
                    } else {
                        this.mChange = true;
                        super.onTouchEvent(MotionEvent.obtain(0, 0, 3, 0.0f, 0.0f, 0));
                        return false;
                    }
                } else if (!(actionMasked == 3 || actionMasked == 5 || actionMasked == 6)) {
                    return false;
                }
            }
            stopNestedScroll();
            return super.onTouchEvent(motionEvent);
        }
        this.mChange = false;
        this.mLastMotionY = y;
        startNestedScroll(2);
        return super.onTouchEvent(motionEvent);
    }

    public void setNestedScrollingEnabled(boolean z) {
        this.mChildHelper.setNestedScrollingEnabled(z);
    }

    public boolean isNestedScrollingEnabled() {
        return this.mChildHelper.isNestedScrollingEnabled();
    }

    public boolean startNestedScroll(int i) {
        return this.mChildHelper.startNestedScroll(i);
    }

    public void stopNestedScroll() {
        this.mChildHelper.stopNestedScroll();
    }

    public boolean hasNestedScrollingParent() {
        return this.mChildHelper.hasNestedScrollingParent();
    }

    public boolean dispatchNestedScroll(int i, int i2, int i3, int i4, int[] iArr) {
        return this.mChildHelper.dispatchNestedScroll(i, i2, i3, i4, iArr);
    }

    public boolean dispatchNestedPreScroll(int i, int i2, int[] iArr, int[] iArr2) {
        return this.mChildHelper.dispatchNestedPreScroll(i, i2, iArr, iArr2);
    }

    public boolean dispatchNestedFling(float f, float f2, boolean z) {
        return this.mChildHelper.dispatchNestedFling(f, f2, z);
    }

    public boolean dispatchNestedPreFling(float f, float f2) {
        return this.mChildHelper.dispatchNestedPreFling(f, f2);
    }
}

File: decompiled/sources/com/hjq/widget/layout/NestedScrollWebView.kt
package com.hjq.widget.layout

import android.content.Context
import android.util.AttributeSet
import android.view.MotionEvent
import android.webkit.WebView
import androidx.core.view.NestedScrollingChild
import androidx.core.view.NestedScrollingChildHelper
import kotlin.math.max

open class NestedScrollWebView : WebView, NestedScrollingChild {
    private var mChange = false
    private val mChildHelper = NestedScrollingChildHelper(this)
    private var mLastMotionY = 0
    private var mNestedOffsetY = 0
    private val mScrollConsumed = IntArray(2)
    private val mScrollOffset = IntArray(2)

    constructor(context: Context) : super(context) {
        isNestedScrollingEnabled = true
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        isNestedScrollingEnabled = true
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    ) {
        isNestedScrollingEnabled = true
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int) : super(
        context,
        attrs,
        defStyleAttr,
        defStyleRes
    ) {
        isNestedScrollingEnabled = true
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        val obtain = MotionEvent.obtain(event)
        val actionMasked = event.actionMasked
        var handled = false
        if (actionMasked == MotionEvent.ACTION_DOWN) {
            mNestedOffsetY = 0
        }
        val y = event.y.toInt()
        event.offsetLocation(0f, mNestedOffsetY.toFloat())
        if (actionMasked != MotionEvent.ACTION_DOWN) {
            if (actionMasked != MotionEvent.ACTION_UP) {
                if (actionMasked == MotionEvent.ACTION_MOVE) {
                    var deltaY = mLastMotionY - y
                    if (dispatchNestedPreScroll(0, deltaY, mScrollConsumed, mScrollOffset)) {
                        deltaY -= mScrollConsumed[1]
                        obtain.offsetLocation(0f, mScrollOffset[1].toFloat())
                        mNestedOffsetY += mScrollOffset[1]
                    }
                    mLastMotionY = y - mScrollOffset[1]
                    val scrollY = scrollY
                    val consumedY = max(0, scrollY + deltaY) - scrollY
                    if (dispatchNestedScroll(0, consumedY, 0, deltaY - consumedY, mScrollOffset)) {
                        mLastMotionY -= mScrollOffset[1]
                        obtain.offsetLocation(0f, mScrollOffset[1].toFloat())
                        mNestedOffsetY += mScrollOffset[1]
                    }
                    if (mScrollConsumed[1] == 0 && mScrollOffset[1] == 0) {
                        if (mChange) {
                            mChange = false
                            obtain.action = MotionEvent.ACTION_DOWN
                            super.onTouchEvent(obtain)
                        } else {
                            handled = super.onTouchEvent(obtain)
                        }
                        obtain.recycle()
                        return handled
                    }
                    if (mChange) {
                        return false
                    }
                    mChange = true
                    super.onTouchEvent(MotionEvent.obtain(0, 0, MotionEvent.ACTION_CANCEL, 0f, 0f, 0))
                    return false
                }
                if (
                    actionMasked != MotionEvent.ACTION_CANCEL &&
                    actionMasked != MotionEvent.ACTION_POINTER_DOWN &&
                    actionMasked != MotionEvent.ACTION_POINTER_UP
                ) {
                    return false
                }
            }
            stopNestedScroll()
            return super.onTouchEvent(event)
        }
        mChange = false
        mLastMotionY = y
        startNestedScroll(2)
        return super.onTouchEvent(event)
    }

    override fun setNestedScrollingEnabled(enabled: Boolean) {
        mChildHelper.isNestedScrollingEnabled = enabled
    }

    override fun isNestedScrollingEnabled(): Boolean {
        return mChildHelper.isNestedScrollingEnabled
    }

    override fun startNestedScroll(axes: Int): Boolean {
        return mChildHelper.startNestedScroll(axes)
    }

    override fun stopNestedScroll() {
        mChildHelper.stopNestedScroll()
    }

    override fun hasNestedScrollingParent(): Boolean {
        return mChildHelper.hasNestedScrollingParent()
    }

    override fun dispatchNestedScroll(
        dxConsumed: Int,
        dyConsumed: Int,
        dxUnconsumed: Int,
        dyUnconsumed: Int,
        offsetInWindow: IntArray?
    ): Boolean {
        return mChildHelper.dispatchNestedScroll(
            dxConsumed,
            dyConsumed,
            dxUnconsumed,
            dyUnconsumed,
            offsetInWindow
        )
    }

    override fun dispatchNestedPreScroll(
        dx: Int,
        dy: Int,
        consumed: IntArray?,
        offsetInWindow: IntArray?
    ): Boolean {
        return mChildHelper.dispatchNestedPreScroll(dx, dy, consumed, offsetInWindow)
    }

    override fun dispatchNestedFling(velocityX: Float, velocityY: Float, consumed: Boolean): Boolean {
        return mChildHelper.dispatchNestedFling(velocityX, velocityY, consumed)
    }

    override fun dispatchNestedPreFling(velocityX: Float, velocityY: Float): Boolean {
        return mChildHelper.dispatchNestedPreFling(velocityX, velocityY)
    }
}
